tests/test_end_to_end.py

```diff
@@ -0,0 +1,44 @@
+import apache_beam as beam
+import pytest
+import xarray as xr
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.testing.test_pipeline import TestPipeline
+
+from pangeo_forge_recipes.patterns import CombineOp, DimKey
+from pangeo_forge_recipes.transforms import (
+    DatasetToSchema,
+    DetermineSchema,
+    IndexItems,
+    OpenWithXarray,
+    PrepareZarrTarget,
+    StoreToZarr,
+)
+
+# from apache_beam.testing.util import assert_that, equal_to
+# from apache_beam.testing.util import BeamAssertException, assert_that, is_not_empty
+
+
+@pytest.fixture
+def pipeline():
+    options = PipelineOptions(runtime_type_check=False)
+    with TestPipeline(options=options) as p:
+        yield p
+
+
+def test_xarray_zarr(
+    daily_xarray_dataset, netcdf_local_file_pattern_sequential, pipeline, tmp_target_url
+):
+    pattern = netcdf_local_file_pattern_sequential
+    with pipeline as p:
+        inputs = p | beam.Create(pattern.items())
+        datasets = inputs | OpenWithXarray(file_type=pattern.file_type)
+        # TODO determine this dynamically
+        combine_dims = [DimKey("time", operation=CombineOp.CONCAT)]
+        schemas = datasets | DatasetToSchema()
+        schema = schemas | DetermineSchema(combine_dims=combine_dims)
+        indexed_datasets = datasets | IndexItems(schema=schema)
+        target = schema | PrepareZarrTarget(target_url=tmp_target_url)
+        _ = indexed_datasets | StoreToZarr(target_store=target)
+
+    ds = xr.open_dataset(tmp_target_url, engine="zarr").load()
+    xr.testing.assert_equal(ds, daily_xarray_dataset)
```

Comment:
This block is really the highlight of this PR. It is the closest we have come yet to recreating the original monolithic XarrayZarrRecipe in modular Beam style.

But there is still a lot of room to refine this. Perhaps some of the intermediate steps can be merged into a single PTransform. I'm not sure the user needs to be passing schemas around.

Really interesting in getting feedback on the API and flow here - @cisaacstern and @alxmrs.

Reply:
I'll review this PR in more depth later today or tomorrow. It's exciting to see all this progress!